### DISCUSSION TITLE
Is there anything wrong with having a large number of providers all initialized and running at the same time?

### DISCUSSION THREAD

**Discussion by mark8044** [rrousselGit/riverpod#1258] [General]
As my app has grown, I now about 45 providers, mostly ChangeNotifierProviders running.  They all get initialized right at main app startup, and continue to run and do various things all throughout the app.

Everyhting is running smoothly and without a problem.

My question is I guess theoretical, since everything is OK.  But the question is, is there anything inherently wrong with having a lot of providers all come up at the same time and run at the same time.  Or maybe 45 isn't considered a lot ...

**Comment by rrousselGit** [comment]
There's nothing wrong with a large number of active providers

It's more about whether your architecture make sense, but that's a different topic.

**Comment by venkatd** [comment]
@mark8044 are all of those 45 providers needed right away?

There's nothing wrong with having so many providers, but if they aren't needed right away, you can take advantage of the lazy instantiation that riverpod provides.

For example, we have a `timeTrackingServiceProvider` and it's only instantiated when a user opens up the popup related to time tracking because the `ref.watch` is scoped to the build method where it's absolutely needed.

**Reply by mark8044** [reply]
No they aren't all needed immediatly at start, maybe about half of them.  

But if there's nothing wrong with a large # of providers as @rrousselGit has mentioned above, then why change it. 

Why fix it if it aint broke :)

**Reply by venkatd** [reply]
Just brainstorming best practice. If it's working for you, no need to change it ;)

Some advantages I see of scoping providers as locally as possible:
- The code can be easier to reason about. We only have a few top-level providers in our app: push notifications, routing, authentication, etc. When I look at the code it's obvious what are dependencies needed by the whole app and which ones are needed by a sub-part of the app.
- Load time. It might be fine for now, but it can add up. If these dependencies call the OS, send API requests, or do some small bit of processing at the start, it can make your UI unresponsive right after rendering the first frame.
